### PR TITLE
Fix crash in Jsonrpc

### PR DESCRIPTION
**Pull request by Paxxi**
## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/interfaces/json-rpc/JSONServiceDescription.cpp [hunk 1]
```diff
@@ -1224,9 +1221,31 @@ void JSONSchemaTypeDefinition::Print(bool isParameter,
...
 {
-  if (typeDefinition.get() == NULL)
+  // Check and set the reference type before recursing
+  // to guard against cycles
+  if (referencedTypeSet)
+    return;
+
+  referencedTypeSet = true;
+
+	// Take care of all nested types
```
Comment (a1rwulf) [inline xbmc/interfaces/json-rpc/JSONServiceDescription.cpp]: Please align comment.

File: xbmc/interfaces/json-rpc/JSONServiceDescription.cpp [hunk 2]
```diff
@@ -1254,6 +1273,8 @@ void JSONSchemaTypeDefinition::Set(const JSONSchemaTypeDefinitionPtr typeDefinit
   if (referencedTypeDef.get() != NULL)
     referencedType = referencedTypeDef;
 
+  // This will have been overwritten by the copy of the refenrece
```
Comment (a1rwulf) [inline xbmc/interfaces/json-rpc/JSONServiceDescription.cpp]: `reference`

**Review by a1rwulf — CHANGES_REQUESTED** [review]
2 nitpicks - looks good otherwise.
Thanks for this!

**Comment by DaveTBlake** [comment]
I must have missed something, but what crash is this fixing?
Looks like a lot of whitespace and format changes, is there something else in there I missed?
Does this not also merit a patch bump to indicate to API consumers that something internal has changed? Or maybe not if it is just cosmetics, but the "crash" in the titile implies it is something more, hence show that this patch version of API does not crash.

**Comment by a1rwulf** [comment]
> I must have missed something, but what crash is this fixing?
> Looks like a lot of whitespace and format changes, is there something else in there I missed?
> Does this not also merit a patch bump to indicate to API consumers that something internal has changed? Or maybe not if it is just cosmetics, but the "crash" in the titile implies it is something more, hence show that this patch version of API does not crash.

The first commit changes codestyle according to clang-format rules, the second commit is the actual fix - so check out the second commit.
Imho we don't need to bump the version for this.

**Comment by Paxxi** [comment]
No behavior should change with this fix. All it does is move the work to be done during init instead of on demand to avoid threading issues.

Visual studio used the clang format so I moved that into its own commit? Might as well get it over with 😀

**Comment by DaveTBlake** [comment]
Ah, I missed the meat for the clang. :)

Does the fix relate to this threading issue https://github.com/xbmc/xbmc/pull/9986 that was discussed by Tolriq and @Montellese but never fixed? Would be good if it does. Although I guess it came out of the discussion of a crash in testing of Cmake C14 PR https://github.com/xbmc/xbmc/pull/13034 is that right?

>No behavior should change with this fix.
OK, but changing how/when the loading is done _could_ have unexpected impact, especially given the earlier comments from @Montellese, so does need some testing right?

It fixes an issue in JSON - the API after this change will not behave the same as before it (otherwise it isn't a fix) - and changes the internal implementation therefore the API patch version needs to be bumped. 

@a1rwulf when to bump the version is not about opinion, it is part of the version number definition in the schema. - "Bumped on any changes to the internal implementation but not to the API definition"
https://github.com/xbmc/xbmc/blob/90c7746ad29c99e20beaaa6f3b4e107d59651890/xbmc/interfaces/json-rpc/schema/methods.json#L34

**Comment by Paxxi** [comment]
Not crashing randomly when an api is called is not something that requires the client to change their behavior so it doesn't need an api bump

**Comment by Rechi** [comment]
I would prefer if you drop the clangformat commit and instead run git-clang-format or clang-format-diff on the real code changes.

**Comment by DaveTBlake** [comment]
@Paxxi patch number version bumps are not about things that require API consumer to change use, just an indicator that internal implementation has changed.   As the schema says - patch number is  "Bumped on any changes to the _internal implementation_ but not to the API definition"

This certainly changes the internal implementation.

**Comment by Paxxi** [comment]
ok remove clang-format, bumped api patch version and fixed the comments

**Review by a1rwulf — APPROVED** [review]

**Comment by a1rwulf** [comment]
jenkins build this please

**Comment by Paxxi** [comment]
@Rechi @DaveTBlake good to go?

**Comment by Rechi** [comment]
> I would prefer if you drop the clangformat commit and instead run git-clang-format or clang-format-diff on the real code changes.

You only did the first part.

**Comment by DaveTBlake** [comment]
Yes @Paxxi, thanks.
In another life I would like to have a set of JSON tests we could run covering all the methods and possible returns, and multiple calls at the same time (from different clients etc.).  But in this one at least we are nice an early in the dev cycle.

**Comment by Rechi** [comment]
wrong revision was build
jenkins build this please

**Review by Rechi — APPROVED** [review]

**Comment by fritsch** [comment]
Thanks much @Paxxi !

**Pull request merged**